### ISSUE TITLE
ENH: Reimplement most of `random` using the ufunc machinery

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#8833]
This would:
 * Add broadcasting (#6938)
 * Add an `out` argument
 * Add a `where` argument
 * Make the output dtype derive from the inputs (eg randint)

Issues:
* right now we implement most of this in cython, and there is no cython header for ufuncs.
   * Are we gaining much from the cython, or is it mostly there to handle the iteration that ufuncs would deal with?
* random functions are bound methods, not functions. Are method ufuncs something that is possible?

**Comment by mhvk** [comment]
I have not looked at the code at all, so my comment may be completely off the mark, but I guess if you think of this as a  `()->()` ufunc, where the input happens to be a new base random number whenever you access it, then it must gain a (temporary) `shape` attribute. So perhaps the method would be do something like make a separate entity of `self` in which it can set the shape (and possibly have all-zero strides), and then call the `ufunc`. It does seem that for this purpose it makes most sense if `self` is an object.

**Comment by bashtage** [comment]
> Make the output dtype derive from the inputs (eg randint)

This one is possibly harder.  `randint` is efficient in the sense that it uses as few bits as are required to produce a random `(u)int*`.  Providing efficient routines for other dypes for most distributions would be a large undertaking.  Of course one could always case float64 -> float32.  I'm also not sure how complex dtypes would be handled since 2-dimensional random number generation is not always simple from a 1-d generator.

**Comment by eric-wieser** [comment]
>  I'm also not sure how complex dtypes would be handled

Just because we make it a ufunc doesn't mean it has to implement a loop for every dtype :). I agree, it's difficult to define a `random` function for `complex`, primarily because the range is harder to specify

**Comment by NeilGirdhar** [comment]
FYI: regarding complex-typed random functions, I'm just in the middle of adding the complex multivariate normal distribution to scipy https://github.com/scipy/scipy/issues/7212

**Comment by NeilGirdhar** [comment]
FYI: regarding complex-typed random functions, I'm just in the middle of adding the complex multivariate normal distribution to scipy https://github.com/scipy/scipy/issues/7212.  I just started yesterday, but I'm having trouble dealing with the complex parameters and outputs and their infrastructure.

**Comment by NeilGirdhar** [comment]
> Make the output dtype derive from the inputs (eg randint)

If the output derives from the inputs, it may not be the input type exactly.  For example, someone could pass a `float64` matrix for the covariance of a complex Gaussian, but they still need to get a `complex128` out.

**Comment by charris** [comment]
Note that instancing `RandomState` allows you to carry around your own random stream. The current functions are all references to methods of a default instance of `RandomState` instantiated when the module is loaded. I don't think it would be easy to duplicate that in a ufunc with the current arguments.

**Comment by eric-wieser** [comment]
> I don't think it would be easy to duplicate that in a ufunc with the current arguments.

Right, so the goal is to make `np.random.RandomState().randint` work - that's what my mention of a ufunc object arg above is

**Comment by eric-wieser** [comment]
> For example, someone could pass a float64 matrix for the covariance of a complex Gaussian, but they still need to get a complex128 out.

This is why `ufuncs` have a `dtype` parameter

**Comment by charris** [comment]
Why is this a better idea than what we have?

**Comment by eric-wieser** [comment]
Because of the advantages I listed in the opening comment.

I guess the last point would be addressed by making the default dtype `np.result_type(low, high)` rather than `int`
